sql/core/src/test/scala/org/apache/spark/sql/sources/v2/TestInMemoryTableCatalog.scala

```diff
@@ -196,7 +196,103 @@ private class InMemoryTable(
   }
 }
 
-private class BufferedRows extends WriterCommitMessage with InputPartition with Serializable {
+object TestInMemoryTableCatalog {
+  val SIMULATE_FAILED_WRITE_OPTION = "spark.sql.test.simulateFailedWrite"
+  val SIMULATE_FAILED_CREATE_PROPERTY = "spark.sql.test.simulateFailedCreate"
+
+  def maybeSimulateFailedTableCreation(tableProperties: util.Map[String, String]): Unit = {
+    if (tableProperties.containsKey(TestInMemoryTableCatalog.SIMULATE_FAILED_CREATE_PROPERTY)
+      && tableProperties.get(TestInMemoryTableCatalog.SIMULATE_FAILED_CREATE_PROPERTY)
+      .equalsIgnoreCase("true")) {
+      throw new IllegalStateException("Manual create table failure.")
+    }
+  }
+
+  def maybeSimulateFailedTableWrite(tableOptions: CaseInsensitiveStringMap): Unit = {
+    val shouldSimulateFailedWrite = tableOptions
+      .getBoolean(TestInMemoryTableCatalog.SIMULATE_FAILED_WRITE_OPTION, false)
+    if (shouldSimulateFailedWrite) {
+      throw new IllegalStateException("Manual write to table failure.")
+    }
+  }
+}
+
+class TestStagingInMemoryCatalog
+  extends TestInMemoryTableCatalog with StagingTableCatalog {
+
+  override def stageCreate(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String]): StagedTable = {
+    newStagedTable(ident, schema, partitions, properties, replaceIfExists = false)
+  }
+
+  override def stageReplace(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String]): StagedTable = {
+    newStagedTable(ident, schema, partitions, properties, replaceIfExists = true)
+  }
+
+  private def newStagedTable(
+      ident: Identifier,
+      schema: StructType,
+      partitions: Array[Transform],
+      properties: util.Map[String, String],
+      replaceIfExists: Boolean): StagedTable = {
+    import CatalogV2Implicits.IdentifierHelper
+    if (partitions.nonEmpty) {
+      throw new UnsupportedOperationException(
+        s"Catalog $name: Partitioned tables are not supported")
+    }
+
+    TestInMemoryTableCatalog.maybeSimulateFailedTableCreation(properties)
+
+    new TestStagedTable(
+      ident,
+      new InMemoryTable(s"$name.${ident.quoted}", schema, properties),
+      replaceIfExists)
+  }
+
+  private class TestStagedTable(
+      ident: Identifier,
+      delegateTable: InMemoryTable,
+      replaceIfExists: Boolean)
+    extends StagedTable with SupportsWrite with SupportsRead {
+
+    override def commitStagedChanges(): Unit = {
+      if (replaceIfExists) {
+        tables.put(ident, delegateTable)
```

Comment:
nit: when committing REPLACE TABLE, we should fail if the table is already dropped by others.